cli/analyze/analyze.go

```diff
@@ -0,0 +1,559 @@
+	longestPathLength := 0
+	var longestPathEnd *string
+
+	length := map[string]int{}
+	pred := map[string]*string{}
+	e := g.EdgeSet()
+	for _, m := range g.TopologicalSort() {
+		for n := range e.Incoming[m] {
+			l := length[n] + 1
+			if l > length[m] {
+				length[m] = l
+				n := n
+				pred[m] = &n
+			}
+			if l > longestPathLength {
+				longestPathLength = l
+				m := m
+				longestPathEnd = &m
+			}
+		}
+	}
+	if longestPathEnd == nil {
+		return nil
+	}
+	path := []string{*longestPathEnd}
+	for {
+		p := pred[path[len(path)-1]]
+		if p == nil {
+			break
+		}
+		path = append(path, *p)
+	}
+	reverseSlice(path)
+	return path
+}
+
+// Roots returns all labels in the graph which have no direct dependents.
+func (g *DependencyGraph) Roots() []string {
+	var roots []string
+	for name := range g.Rules {
+		if len(g.DirectDependents[name]) == 0 {
+			roots = append(roots, name)
+		}
+	}
+	return roots
+}
+
+func (g *DependencyGraph) EdgeSet() *EdgeSet {
+	outgoing := map[string]map[string]bool{}
+	incoming := map[string]map[string]bool{}
+	s := g.Roots()
+	for len(s) > 0 {
+		n := s[len(s)-1]
+		s = s[:len(s)-1]
+		if _, ok := outgoing[n]; ok {
+			continue // already visited
+		}
+		rule := g.Rules[n]
+		if rule == nil {
+			continue
+		}
+		for _, m := range rule.RuleInput {
+			if outgoing[n] == nil {
+				outgoing[n] = map[string]bool{}
+			}
+			if incoming[m] == nil {
+				incoming[m] = map[string]bool{}
+			}
+			outgoing[n][m] = true
+			incoming[m][n] = true
+
+			s = append(s, m)
+		}
+	}
+	return &EdgeSet{Outgoing: outgoing, Incoming: incoming}
+}
+
+// TopologicalSort returns all node labels in the graph sorted so that each node
+// is sorted before any of its transitive deps.
+func (g *DependencyGraph) TopologicalSort() []string {
+	// https://en.wikipedia.org/wiki/Topological_sorting#Kahn's_algorithm
+	var out []string
+	s := g.Roots()
+	e := g.EdgeSet()
+	for len(s) > 0 {
+		n := s[len(s)-1]
+		s = s[:len(s)-1]
+		out = append(out, n)
+		for m := anyKey(e.Outgoing[n]); m != nil; m = anyKey(e.Outgoing[n]) {
+			e.Remove(n, *m)
+			if len(e.Incoming[*m]) == 0 {
+				s = append(s, *m)
+			}
+		}
+	}
+	return out
+}
+
+func getAllSourceFilePaths(dir string) (map[string]bool, error) {
+	lines, err := getOutputLines(dir, "git", "ls-files", "--exclude-standard")
+	if err != nil {
+		return nil, fmt.Errorf("git ls-files failed: %s", err)
+	}
+	return makeSet(lines), nil
+}
+
+func getOutputLines(dir, executable string, args ...string) ([]string, error) {
+	cmd := exec.Command(executable, args...)
+	cmd.Dir = dir
+	stdout := &bytes.Buffer{}
+	stderr := &bytes.Buffer{}
+	cmd.Stdout = stdout
+	cmd.Stderr = stderr
+	if err := cmd.Run(); err != nil {
+		return nil, fmt.Errorf("%s", strings.TrimSpace(stderr.String()))
+	}
+	output := strings.TrimRight(stdout.String(), "\n")
+	if output == "" {
+		return nil, nil
+	}
+	return strings.Split(output, "\n"), nil
+}
+
+func anyKey[K comparable, V any](m map[K]V) *K {
```

Comment:
Could you add comments to the functions here and below?

Reply:
Done